### PR TITLE
RL Agent

### PR DESCRIPTION
**Pull request by NoB0**
Add agent and dialogue manager updated for RL.

### CODE REVIEW

File: reinforcement_learning/agent/rl_agent.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        )
+
+        if config.get("nlu_type", "") == "neural":
+            self.nlu = NeuralNLU(None)
+
+    def initialize(self) -> None:
+        """Initializes the agent."""
+        self.dialogue_manager.initialize()
+
+    def generate_utterance(
```
Comment (IKostric) [inline reinforcement_learning/agent/rl_agent.py:40]: I am not sure how this method is being used. It seems to me it is a mix of `recieve_utterance` and `_generate_utterance.` We need to have `recieve_utterance` defined for the integration to work.

File: reinforcement_learning/agent/rl_agent.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        self,
+        agent_dacts: List[DialogueAct],
+        options: DialogueOptions = {},
+        user_fname: str = None,
```
Comment (IKostric) [inline reinforcement_learning/agent/rl_agent.py:44]: Why is `user_fname` added here? I presume this is needed for NLG to output custom text. If this is the case, It should be a part of agent DialogueAct (HI), maybe retrieved from UserModel?

File: reinforcement_learning/agent/rl_agent.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        Returns:
+            List of dialogue acts.
+        """
+        return self.nlu.generate_dacts(
+            user_utterance,
+            utterance_options,
+            self.dialogue_manager.get_state(),
+        )
+
+    def update_placeholder_dialogue_act(
```
Comment (IKostric) [inline reinforcement_learning/agent/rl_agent.py:118]: Nit: The purpose of this method is not clear. Maybe add some more information to the docstrings.

File: reinforcement_learning/agent/rl_agent.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        """Initializes the agent."""
+        self.dialogue_manager.initialize()
+
+    def generate_utterance(
+        self,
+        agent_dacts: List[DialogueAct],
+        options: DialogueOptions = {},
+        user_fname: str = None,
+        recommended_item: Dict[str, Any] = None,
+    ):
```
Comment (IKostric) [inline reinforcement_learning/agent/rl_agent.py:46]: Nit: Missing return type.

**Review by IKostric — CHANGES_REQUESTED** [review]
There are some missing bits to get a proper understanding of the design choices. 
Main questions are:
  - Why do we need completely separate agent if the focus is on DP?
  - Is this PR for training DP only or inference as well?

File: reinforcement_learning/agent/rl_agent.py (hunk 3 above)
Comment (NoB0) [reply reinforcement_learning/agent/rl_agent.py:118]: Good point, I removed the method and updated the docstring of `get_filled_dialogue_acts`. Please check that the purpose is clear now.

File: reinforcement_learning/agent/rl_agent.py (hunk 2 above)
Comment (NoB0) [reply reinforcement_learning/agent/rl_agent.py:44]: As assumed, `user_fname` is needed for NLG. While you suggestion to add it as port of the diaolgue act HI, it should be part of another PR related to the improvement of NLG.

File: reinforcement_learning/agent/rl_agent.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                intent=agent_intent,
+                text=agent_response,
+                participant=DialogueParticipant.AGENT,
+                annotations=[],
+                metadata=metadata,
+            )
+        else:
+            record_data = self.dialogue_manager.get_state().to_dict()
+            metadata.update({"record_data": record_data})
+            utterance = AnnotatedUtterance(
```
Comment (IKostric) [inline reinforcement_learning/agent/rl_agent.py:87]: Code duplication. 
Suggestion: Extract utterance definition outside if/else.

**Review by IKostric — APPROVED** [review]
LGTM, just minor nit

File: reinforcement_learning/agent/rl_agent.py (hunk 5 above)
Comment (NoB0) [reply reinforcement_learning/agent/rl_agent.py:87]: Done.

**Pull request merged**